### PR TITLE
Halo2 SHPLONK benchmark

### PR DESCRIPTION
**Pull request by tyshko-rostyslav**

### CODE REVIEW

File: halo2-shplonk/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,387 @@
+use ark_std::{end_timer, start_timer};
+use ff::Field;
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:2]: Unused.

File: halo2-shplonk/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+use halo2_curves::bn256::Fr;
+use halo2_proofs::circuit::Value;
+use halo2_proofs::halo2curves as halo2_curves;
+use halo2_proofs::plonk::Circuit;
+use halo2_proofs::{halo2curves::bn256::Bn256, poly::kzg::commitment::ParamsKZG};
+use pasta_curves::{pallas};
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:8]: Unused.

File: halo2-shplonk/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+use rand::RngCore;
+use rand::rngs::OsRng;
+use snark_verifier_sdk::evm::{evm_verify, gen_evm_proof_shplonk, gen_evm_verifier_shplonk};
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:11]: Unused.

File: halo2-shplonk/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+        fn secure_mds() -> usize {
+            0
+        }
+
+        fn constants() -> (Vec<[Fr; WIDTH]>, Mds<Fr, WIDTH>, Mds<Fr, WIDTH>) {
+            generate_constants::<_, Self, WIDTH, RATE>()
+        }
+    }
+
+    const K: u32 = 7;
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:188]: Unused.

File: halo2-shplonk/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+                    a.copy_advice(|| "", &mut region, config.c, 4)?;
+
+                    Ok(())
+                },
+            )
+        }
+    }
+}
+
+fn gen_application_snark_old(params: &ParamsKZG<Bn256>) -> Snark {
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:328]: This function is unused.

File: halo2-shplonk/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+        &params,
+        &agg_circuit.without_witnesses(),
+        None,
+        // Some(Path::new("./examples/agg.pk")),
+    );
+    end_timer!(start0);
+
+    println!("after pk");
+
+    let agg_instances = agg_circuit.instance();
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:375]: Unused.

File: halo2-shplonk/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+                instance: Fr::ZERO,
+                message_arr: [Fr::ZERO; L],
+                message: Value::unknown(),
+                _spec: PhantomData,
+            }
+        }
+
+        fn configure(meta: &mut ConstraintSystem<Fr>) -> Self::Config {
+            let state = (0..WIDTH).map(|_| meta.advice_column()).collect::<Vec<_>>();
+            let expected = meta.instance_column();
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:103]: Unused.

File: halo2-shplonk/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+                    let message: Result<Vec<_>, Error> = (0..L).map(message_word).collect();
+                    Ok(message?.try_into().unwrap())
+                },
+            )?;
+
+            let hasher = Hash::<_, _, S, ConstantLength<L>, WIDTH, RATE>::init(
+                chip,
+                layouter.namespace(|| "init"),
+            )?;
+            let output = hasher.hash(layouter.namespace(|| "hash"), message)?;
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:154]: Unused.

Comment (jonesmarvin8) [reply halo2-shplonk/src/main.rs:154]: Note that the removal of this line causes an error on line 146. So, even though the variable output is never used, I think this line provides the compiler with information concerning variable message.

File: halo2-shplonk/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+
+    use halo2_gadgets::poseidon::{
+        primitives::{self as poseidon, generate_constants, ConstantLength, Mds, Spec},
+        Hash, Pow5Chip, Pow5Config,
+    };
+    use std::convert::TryInto;
+    use std::marker::PhantomData;
+
+    use criterion::{criterion_group, criterion_main, Criterion};
+    use rand::rngs::OsRng;
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:62]: Unused.

File: halo2-shplonk/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+    use halo2curves::pasta::{pallas, vesta, EqAffine};
+
+    use halo2_gadgets::poseidon::{
+        primitives::{self as poseidon, generate_constants, ConstantLength, Mds, Spec},
+        Hash, Pow5Chip, Pow5Config,
+    };
+    use std::convert::TryInto;
+    use std::marker::PhantomData;
+
+    use criterion::{criterion_group, criterion_main, Criterion};
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:61]: Unused.

File: halo2-shplonk/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+                multiopen::ProverIPA,
+                strategy::SingleStrategy,
+            },
+            VerificationStrategy,
+        },
+        transcript::{
+            Blake2bRead, Blake2bWrite, Challenge255, TranscriptReadBuffer, TranscriptWriterBuffer,
+        },
+    };
+    use halo2curves::pasta::{pallas, vesta, EqAffine};
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:52]: Unused.

File: halo2-shplonk/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+        poly::{
+            commitment::ParamsProver,
+            ipa::{
+                commitment::{IPACommitmentScheme, ParamsIPA},
+                multiopen::ProverIPA,
+                strategy::SingleStrategy,
+            },
+            VerificationStrategy,
+        },
+        transcript::{
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:48]: All of these transcript dependencies are unused.

File: halo2-shplonk/src/main.rs (hunk 5 above)
Comment (tyshko-rostyslav) [reply halo2-shplonk/src/main.rs]: Removed

File: halo2-shplonk/src/main.rs (hunk 8 above)
Comment (tyshko-rostyslav) [reply halo2-shplonk/src/main.rs]: We cannot remove it as it doing the main job of the circuit - hashing

File: halo2-shplonk/src/main.rs (hunk 4 above)
Comment (tyshko-rostyslav) [reply halo2-shplonk/src/main.rs]: Removed

File: halo2-shplonk/src/main.rs [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+use snark_verifier_sdk::{
+    gen_pk,
+    halo2::{aggregation::AggregationCircuit, gen_snark_shplonk},
+    Snark,
+};
+use snark_verifier_sdk::{CircuitExt, SHPLONK};
+use std::marker::PhantomData;
+use std::path::Path;
+
+pub mod application {
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:19]: Lines 19-293 is mostly modified PSE code. Could this be shifted to a separate file (with credit provided)? As mentioned on halo2-gwc code, I'm not sure if we can use cargo to pull dependencies from a snark-verifier-sdk/example.

File: halo2-shplonk/src/main.rs [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+        pub instance: Fr,
+        pub message_arr: [Fr; L],
+        pub message: Value<[Fr; L]>,
+        pub _spec: PhantomData<S>,
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct MyConfig<const WIDTH: usize, const RATE: usize, const L: usize> {
+        input: [Column<Advice>; L],
+        // expected: Column<Instance>,
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:51]: Unused.

File: halo2-shplonk/src/main.rs [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+
+    let params = gen_srs(23);
+    let agg_circuit = AggregationCircuit::<SHPLONK>::new(&params, snarks);
+    println!("after agg_circuit");
+    let start0 = start_timer!(|| "gen vk & pk");
+    let pk = gen_pk(
+        &params,
+        &agg_circuit.without_witnesses(),
+        None,
+        // Some(Path::new("./examples/agg.pk")),
```
Comment (jonesmarvin8) [inline halo2-shplonk/src/main.rs:327]: Unused.

File: halo2-shplonk/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+rand = "0.8"
+rand_chacha = "0.3.1"
+hex = "0.4"
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0"
+bincode = "1.3.3"
+# ark-std = { version = "0.3.0", features = ["print-trace"], optional = true }
+
+halo2_gadgets = { git = "https://github.com/privacy-scaling-explorations/halo2", tag = "v2023_04_20" } # not optional for now
+pasta_curves = "0.5"
```
Comment (jonesmarvin8) [inline halo2-shplonk/Cargo.toml:23]: Unused.

File: halo2-shplonk/Cargo.toml [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+# loader_evm
+ethereum-types = { version = "0.14", default-features = false, features = [
+    "std",
+], optional = true }
+
+ark-std = { version = "0.3.0", features = ["print-trace"] }
+paste = "1.0.7"
+pprof = { version = "0.11", features = ["criterion", "flamegraph"] }
+criterion = "0.4"
```
Comment (jonesmarvin8) [inline halo2-shplonk/Cargo.toml:45]: Unused.

File: halo2-shplonk/Cargo.toml [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+criterion-macro = "0.4"
```
Comment (jonesmarvin8) [inline halo2-shplonk/Cargo.toml:46]: Unused.

File: halo2-shplonk/Cargo.toml [hunk 19]
```diff
@@ -0,0 +1,61 @@
+[package]
+name = "halo2-shplonk"
+version = "0.1.0"
+edition = "2021"
```
Comment (jonesmarvin8) [inline halo2-shplonk/Cargo.toml:4]: Year should be updated.

**Comment by jonesmarvin8** [comment]
Note that I cannot fully execute the code on my machine. So, any suggestion can only be verifier by building the code and running it up to a point.

File: halo2-shplonk/src/main.rs (hunk 13 above)
Comment (tyshko-rostyslav) [reply halo2-shplonk/src/main.rs:19]: Unfortunately in cannot

File: halo2-shplonk/Cargo.toml (hunk 19 above)
Comment (tyshko-rostyslav) [reply halo2-shplonk/Cargo.toml:4]: What do you mean?

File: halo2-shplonk/Cargo.toml [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+itertools = "0.10.3"
+lazy_static = "1.4.0"
+num-bigint = "0.4.3"
+num-integer = "0.1.45"
+num-traits = "0.2.15"
+rand = "0.8"
+rand_chacha = "0.3.1"
```
[the remark this answers is not in the document]
Comment (tyshko-rostyslav) [reply halo2-shplonk/Cargo.toml:15]: `rand` is needed, removed everything else

**Pull request merged**